packages/slate-dev-benchmark/src/components/Bench.js

```diff
@@ -0,0 +1,219 @@
+
+  run(runner) {
+    this.runner = runner
+  }
+
+  /*
+   * Tries to run tasks in `times`, if the time consuming excedes the max-time, then stop;
+   * After run, generate report and return
+   * If initial is the initial index to run the task, for continueing a task in adaptive mode
+   * @param {number} times
+   * @param {number} initial
+  */
+
+  async compose(times, initial) {
+    times = Math.floor(times)
+    const isAsync = this.options.async
+    const { runner, inputer } = this
+
+    const { maxTime } = this.options
+    let seq = Number.isFinite(this.options.maxTries) ? 1 : NaN
+    let nextCheckIndex = seq
+    const hrStart = process.hrtime()
+
+    if (global.gc) {
+      global.gc()
+    }
+
+    const report = { user: 0, system: 0, all: 0, hr: 0, cycles: 0 }
+    for (
+      let initialIndex = initial;
+      initialIndex < times;
+      initialIndex += this.options.allocationTries
+    ) {
+      const tries = Math.min(times - initialIndex, this.options.allocationTries)
+      const thisTryReport = await compposeAnAllocation.call(
+        this,
+        tries,
+        initialIndex
+      )
+      if (global.gc) {
+        global.gc()
+      }
+      for (const key in report) {
+        report[key] += thisTryReport[key]
+      }
+    }
+    return report
+
+    /*  Run the task from initialIndex to index by {tries}
+     *  Split task to avoid allocating too much memories for {input} in the same time
+     *  @param {number} tries
+     *  @param {number} initialIndex
+     *  @return {Promise< Object , *>}
+    */
+
+    function compposeAnAllocation(tries, initialIndex) {
+      const inputs = Array.from({ length: tries }).map(index =>
+        inputer(index + initialIndex)
+      )
+      const timer = new Timer()
+      timer.start()
+      return dispatch(0).then(cycles => {
+        timer.end()
+        const { elapsed } = timer
+        return { ...elapsed, cycles }
+      })
+
+      /*  Run a single task run; If the task is end, return a Promise with the index when the task ends
+     *  @param {number} index
+     *  @return {Promise<number, *>}
+     */
+
+      function dispatch(index) {
+        if (index === tries) return Promise.resolve(tries)
+        if (index === nextCheckIndex) {
+          const hrEnd = process.hrtime(hrStart)
+          const elapsed = hrEnd[0] * 1e3 + hrEnd[1] / 1e6
+          if (elapsed > maxTime) {
+            return Promise.resolve(index)
+          } else {
+            if (elapsed < maxTime / 20) {
+              seq *= 2
+            }
+            nextCheckIndex = seq + nextCheckIndex
+          }
+        }
+        if (!isAsync) {
+          const inputVar = inputs[index]
+          runner(inputVar)
+          return dispatch(index + 1)
+        } else {
+          return Promise.resolve(runner(inputs[index])).then(() =>
+            dispatch(index + 1)
+          )
+        }
+      }
+    }
+  }
+
+  /*
+   * Run the bench
+   * @return {void}
+  */
+
+  makeRun() {
```

Comment:
Because I want to expose Repository.run, Bench.run as the interface.  Then I have to use Bench.makeRun as the inner interface to run benches. 
To be consistent with Bench, I choose makeRun as the inner interface called by Repository.